Initialize

LouisSzeto opened the pull request:
Do we need to make it downloadable? 

or can we make it able to do a single line call by something like
`data = AddData<NasdaqDataLink>("PROVIDER/SYMBOL", "desire_value_column_name", "api_key");`
as right now, the user would need to copy and modify in their project I think

LouisSzeto committed:
Update Demonstration.cs

LouisSzeto committed:
Delete output/alternative/mycustomdatatype directory

LouisSzeto committed:
Update MyCustomDataTypeTests.cs

LouisSzeto committed:
Rename MyCustomDataTypeTests.cs to NasdaqDataLinkTests.cs

LouisSzeto committed:
Update and rename MyCustomDataType.cs to NasdaqDataLink.cs

With reference to https://github.com/QuantConnect/Lean/blob/5d762d16b2ec7bd48fd766b772b0921862deda22/Common/Data/Custom/Quandl.cs

LouisSzeto committed:
Update QuantConnect.DataSource.csproj

LouisSzeto committed:
Delete process.sample.ipynb

LouisSzeto committed:
Delete process.sample.py

LouisSzeto committed:
Delete process.sample.sh

LouisSzeto committed:
Update NasdaqDataLink.cs

LouisSzeto committed:
Create Quandl.cs

LouisSzeto committed:
Update Quandl.cs

LouisSzeto committed:
Update QuantConnect.DataSource.csproj

LouisSzeto committed:
Update NasdaqDataLinkTests.cs

LouisSzeto committed:
Update QuantConnect.DataSource.csproj

LouisSzeto committed:
Update Quandl.cs

LouisSzeto committed:
Update NasdaqDataLink.cs

LouisSzeto committed:
Update NasdaqDataLink.cs

Martin-Molinero commented on QuantConnect.DataSource.csproj:
```diff
@@ -3,6 +3,8 @@
     <TargetFramework>net5.0</TargetFramework>
     <RootNamespace>QuantConnect.DataSource</RootNamespace>
+    <RootNamespace>QuantConnect.DataSource.NasdaqDataLink</RootNamespace>
+    <RootNamespace>QuantConnect.DataSource.Quandl</RootNamespace>
```
double RootNamespace 😛

Martin-Molinero commented on Quandl.cs:
```diff
@@ -0,0 +1,33 @@
+    /// Quandl Data Type (Depreciated, Use NasdaqDataLink instead.)
+    /// </summary>
+    [ObsoleteAttribute("Quandl is obsolete. Use NasdaqDataLink instead.", false)]
+    public class Quandl : NasdaqDataLink
```
Cool! Missing the PR for Lean removing it over there 👍

Martin-Molinero commented on NasdaqDataLink.cs:
```diff
@@ -0,0 +1,194 @@
+            if (commonList.Any())
+            {
+                // If so, set .Value as the first common element with price keyword
+                data.Value = (decimal)data.GetProperty(commonList[0]);
```
Cool! we should still allow users to overwrite/specify this somehow I think, would it be possible? nice unit test to add 👍 

`_valueColumn` is no longer being used BTW, maybe it's optional now, default value null and we try `{"close", "price", "settle", "value"}` and if desired user can specify it

Martin-Molinero commented on NasdaqDataLink.cs:
```diff
@@ -0,0 +1,194 @@
+                    _propertyNames.Add(property);
+                }
+                // Returns null at this point where we are only reading the properties names
+                return null;
```
Why would it return null, what if `line` is a real data point? 🤔

LouisSzeto replied:
This is default of the .csv export format from Quandl/Nasdaq data link. The first line must be the column name. So I guess it's ok in this way?

AlexCatarino commented on NasdaqDataLink.cs:
```diff
@@ -0,0 +1,194 @@
+            }
+
+            data.Time = DateTime.ParseExact(csv[0], "yyyy-MM-dd", CultureInfo.InvariantCulture);
+            data.EndTime = data.Time + TimeSpan.FromDays(1);
```
I think we should use
```csharp
data.EndTime = data.Time + config.Resolution.ToTimeSpan()
```
unless we are 100% sure that all data has daily resolution. Given the [default is Resolution.Daily](https://github.com/QuantConnect/Lean.DataSource.NasdaqDataLink/pull/1/files#diff-a58be19fb84f78ea7c46024420b29dc02a00e57eb039ca0e26cf62976f374564R174), it will likely fall in this case, but we give more options here.

AlexCatarino commented on NasdaqDataLink.cs:
```diff
@@ -0,0 +1,194 @@
+        /// </summary>
+        /// <param name="config">Subscription configuration</param>
+        /// <param name="line">Line of data</param>
+        /// <param name="date">Date</param>
```
nit: We could use the docs from Quandl in Lean when there are more informative.
```csharp
        /// <param name="line">CSV line of data from the souce</param>
        /// <param name="date">Date of the requested line</param>
```

AlexCatarino commented on NasdaqDataLink.cs:
```diff
@@ -0,0 +1,194 @@
+        /// <returns>false</returns>
+        public override bool RequiresMapping()
+        {
+            return false;
```
Default is false, so we can delete this method override.

AlexCatarino commented on NasdaqDataLink.cs:
```diff
@@ -0,0 +1,194 @@
+        /// </summary>
+        public override List<Resolution> SupportedResolutions()
+        {
+            return DailyResolution;
```
I wonder if it doesn't support other resolutions, thus we should use `AllResolutions` here.

LouisSzeto committed:
Update NasdaqDataLinkTests.cs

LouisSzeto committed:
Update NasdaqDataLink.cs

Update addressing comments, adding get and set method to allow users specific Value column. Added unit test for that

LouisSzeto committed:
Update QuantConnect.DataSource.csproj

LouisSzeto committed:
Update NasdaqDataLink.cs

AlexCatarino commented on NasdaqDataLink.cs:
```diff
@@ -57,19 +57,23 @@ public static bool IsAuthCodeSet
+        /// Default NasdaqDataLink constructor uses null as its value column
         /// </summary>
-        /// <param name="valueColumnName"></param>
-        protected NasdaqDataLink(string valueColumnName)
```
Removing this constructor is a breaking change. There is a lot of algorithms using Quandl that have the following construction:
```csharp
        public class QuandlMortgagePriceColumns : Quandl
        {
            public QuandlMortgagePriceColumns()            
                // Rename the Quandl object column to the data we want, which is the 'Value' column
	        // of the CSV that our API call returns
                : base(valueColumnName: "Value")
            {
            }
        }
```
that will break.

LouisSzeto replied:
committed your changes. Thank you!

AlexCatarino reviewed (changes requested):
Thank you for the updates, Louis.
Let's work together tomorrow to move this PR. I think there might be other places to fix here and in the Lean PR.

AlexCatarino committed:
Undo Breaking Changes

Martin-Molinero commented:
Superseded by https://github.com/QuantConnect/Lean.DataSource.NasdaqDataLink/pull/2 (current commits included)
